### PR TITLE
tp-qemu: Add a new case for balloon service testing.

### PR DESCRIPTION
**Pull request by CongSmile**
Signed-off-by: Cong Li <coli@redhat.com>

id: 1231037

### CODE REVIEW

File: qemu/tests/balloon_service.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            for memory in xrange(start, end, step):
+                logging.debug("balloon vm mem to: %s B" % memory)
+                vm.monitor.send_args_cmd("balloon value=%s" % memory)
+                vm.monitor.query("balloon")
+                vm.monitor.send_args_cmd(get_polling_cmd % (device_path,
+                                                            get_balloon_property))
+                logging.debug("balloon vm mem to: %s B" % memory)
+                memory = end - memory
+                vm.monitor.send_args_cmd("balloon value=%s" % memory)
+                current_mem = vm.monitor.query("balloon")
```
Comment (xutian) [inline qemu/tests/balloon_service.py]: where this values used ?

File: qemu/tests/cfg/balloon_service.cfg [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    balloon = balloon0
+    balloon_dev_devid = balloon0
+    balloon_dev_add_bus = yes
+    need_enable_verifier = yes
+    need_clear_verifier = yes
+    cdroms = "cd1 winutils virtiowin"
+    drive_index_virtiowin = 3
+    cdrom_virtiowin = iso/windows/virtio-win-1.8.0-4.iso
+    # balloon service related cmd
+    install_balloon_service = "E:\Balloon\w7\amd64\blnsvr.exe -i"
```
Comment (xutian) [inline qemu/tests/cfg/balloon_service.cfg]: Only win7 support it ?

Comment (xutian) [inline qemu/tests/cfg/balloon_service.cfg]: Better to get driver letter by  python code.  We need to make our code work in PPC and ARM platform.

File: qemu/tests/cfg/balloon_service.cfg [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    uninstall_balloon_service = "E:\Balloon\w7\amd64\blnsvr.exe -u"
+    status_balloon_service = "E:\Balloon\w7\amd64\blnsvr.exe status"
```
Comment (xutian) [inline qemu/tests/cfg/balloon_service.cfg]: Please check with feature owner ,  does this file copy to windows32  folder after install virtio_balloon device driver

File: qemu/tests/cfg/balloon_service.cfg (hunk 2 above)
Comment (CongSmile) [inline qemu/tests/cfg/balloon_service.cfg]: It's my mistake, will update it soon.

Thanks.

File: qemu/tests/balloon_service.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    7) uninstall balloon service and clear driver verifier.
+    """
+
+    def balloon_service(session):
+        """
+        Check / Install balloon service.
+
+        :param session: VM session.
+        """
+        status_balloon_service = params.get("status_balloon_service")
```
Comment (xutian) [inline qemu/tests/balloon_service.py]: make  your logical more clearly. I think it should be
  if not running, check is it installed else install it and check is it running

File: qemu/tests/balloon_service.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    try:
+        error_context.context("Enable %s driver verifier in guest" % driver,
+                              logging.info)
+        if params.get("need_enable_verifier", "yes") == "yes":
+            session = utils_test.qemu.setup_win_driver_verifier(session, driver,
+                                                                vm, timeout)
+
+        error_context.context("Install balloon service in guest", logging.info)
+        balloon_service(session)
```
Comment (xutian) [inline qemu/tests/balloon_service.py]: from step info, it looks you want to install balloon service in guest,  but I can't  understand it from your function name

File: qemu/tests/cfg/balloon_service.cfg (hunk 3 above)
Comment (CongSmile) [inline qemu/tests/cfg/balloon_service.cfg]: No, it won't be copied to windows32 folder after install virtio_balloon driver.

**Comment by CongSmile** [comment]
Hi @xutian 

About the config problem, e.g      install_balloon_service = "%s:\Balloon\w7\amd64\blnsvr.exe -i" 

1. the drive letter is gotten from the python code now.
2. About the path w7/amd64, there are 2 ways to solve this problem:
    1) set the config in every windows guest-os cfg.
    2) use the python code to get it, I think it's a little complex.
What do you think?

Thanks.

File: qemu/tests/balloon_service.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        :param session: VM session.
+        :return: the free memory size inside guest.
+        :rtype: int
+        """
+        logging.debug("Get free memory in guest.")
+        free_mem_cmd = params['free_mem_cmd']
+        try:
+            # In Windows guest we get the free memory for memory compare
+            memory = session.cmd_output(free_mem_cmd)
+            memory = int(re.findall(r"\d+", memory)[0])
```
Comment (spcui) [inline qemu/tests/balloon_service.py]: undefined name 're'

File: qemu/tests/balloon_service.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        :param session: VM session.
+        :param get_polling_output: output of get polling in qmp.
+        """
+        error_context.context("Check whether memory status as expected",
+                              logging.info)
+        ratio = params.get("ratio", "0.1")
+        free_mem_in_guest = get_memory_status(session)
+        # Need translate the memory from MB to KB.
+        free_mem_in_guest = free_mem_in_guest * 1024
+        stat_free_memory_qmp = qmp_dict['return']['stats']['stat-free-memory']
```
Comment (spcui) [inline qemu/tests/balloon_service.py]: undefined name 'qmp_dict'

**Comment by CongSmile** [comment]
@spcui 

I have updated the code, could you help review it again?

Thanks.

File: qemu/tests/balloon_service.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        """
+        Get virtio-win disk volume letter.
+
+        :param session: VM session.
+
+        :return drive_letter: virtio-win disk volume letter.
+        """
+        volume_name = params["cdrom_virtiowin"]
+        volume_name = volume_name.split("/")[-1]
+        volume_name = volume_name.split(".iso")[0]
```
Comment (spcui) [inline qemu/tests/balloon_service.py]: strange(".iso"),
volume_name = volume_name[:volume_name.rfind(".")] will be better.

**Comment by CongSmile** [comment]
Thanks @spcui 

Have updated the code, could you review it again?

File: qemu/tests/balloon_service.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        """
+        Check memory status.
+
+        :param session: VM session.
+        :param get_polling_output: output of get polling in qmp.
+        :param keyname: key name of the output of the 'qom-get' property.
+        """
+        error_context.context("Check whether memory status as expected",
+                              logging.info)
+        ratio = params.get("ratio", "0.1")
```
Comment (spcui) [inline qemu/tests/balloon_service.py]: ratio type error, should be float(params.get("ratio", "0.1"))

Comment (CongSmile) [inline qemu/tests/balloon_service.py]: ratio is float type as default.
Could you help provide more details? I have not met such error in my testing.
Not sure why there is type error.
Thanks.

Comment (spcui) [inline qemu/tests/balloon_service.py]: you sure 'ratio = params.get("ratio", "0.1")' is float type? params.get will return a string.

**Comment by CongSmile** [comment]
My fault that add more "" in 0.1:
params.get("ratio", "0.1")

Have updated it. And I have specified the float type of ratio.
Thanks.

File: qemu/tests/balloon_service.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+        :param get_polling_output: output of get polling in qmp.
+        :param keyname: key name of the output of the 'qom-get' property.
+        """
+        error_context.context("Check whether memory status as expected",
+                              logging.info)
+        ratio = float(params.get("ratio", 0.1))
+        if keyname == "stat-free-memory":
+            free_mem_in_guest = get_memory_status(session)
+            # Need translate the memory from KB to B.
+            free_mem_in_guest = free_mem_in_guest * 1024
```
Comment (xutian) [inline qemu/tests/balloon_service.py]: normalize_data_size

**Comment by spcui** [comment]
LGTM, Acked-by: scui@redhat.com

File: qemu/tests/balloon_service.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
+import logging
+import time
+
+from virttest import utils_test
+from virttest import utils_misc
+from virttest import error_context
+from aexpect import ShellCmdError
+
+
+@error_context.context_aware
```
Comment (suqinhuang) [inline qemu/tests/balloon_service.py]: Hi coli,
Suggest to update the case name, no idea what will be tested when I read the case name  balloon_service

File: qemu/tests/balloon_service.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+        :param get_polling_output: output of get polling in qmp.
+        :param keyname: key name of the output of the 'qom-get' property.
+        """
+        error_context.context("Check whether memory status as expected",
+                              logging.info)
+        ratio = float(params.get("ratio", 0.1))
+        if keyname == "stat-free-memory":
+            free_mem_in_guest = utils_test.qemu.get_guest_free_mem(session)
+            # Need translate the memory from KB to B.
+            free_mem_in_guest = free_mem_in_guest * 1024
```
Comment (xutian) [inline qemu/tests/balloon_service.py]: convert type here please.

File: qemu/tests/balloon_service.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+            stat_free_memory_qmp = get_polling_output['return']['stats'][keyname]
+            # The difference of free memory in guest and stat_free_memory in qmp
```
Comment (xutian) [inline qemu/tests/balloon_service.py]: same as above line

File: qemu/tests/balloon_service.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+        :param session: VM session.
+        :param get_polling_output: output of get polling in qmp.
+        :param keyname: key name of the output of the 'qom-get' property.
+        """
+        error_context.context("Check whether memory status as expected",
+                              logging.info)
+        ratio = float(params.get("ratio", 0.1))
+        if keyname == "stat-free-memory":
+            free_mem_in_guest = utils_test.qemu.get_guest_free_mem(session)
```
Comment (xutian) [inline qemu/tests/balloon_service.py]: convert type here, please

File: qemu/tests/balloon_service.py [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+            # Need translate the memory from KB to B.
+            free_mem_in_guest = free_mem_in_guest * 1024
+            stat_free_memory_qmp = get_polling_output['return']['stats'][keyname]
+            # The difference of free memory in guest and stat_free_memory in qmp
+            # should be <= free memory in guest * ratio.
+            if ((abs(long(free_mem_in_guest - long(stat_free_memory_qmp)))) > (long(free_mem_in_guest) * ratio)):
+                raise error_context.TestFail("free memory in guest %s is not"
```
Comment (xutian) [inline qemu/tests/balloon_service.py]: syntax issue

File: qemu/tests/balloon_service.py [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+                                             free_mem_in_guest,
+                                             stat_free_memory_qmp, ratio)
+        elif keyname == "stat-total-memory":
+            guest_mem = params["mem"]
+            # Need translate the memory from MB to B.
+            guest_memory = int(guest_mem) * 1024 * 1024
+            # The difference of total memory of guest and stat_total_memory in
+            # qmp should be <= total memory of guest * ratio.
+            stat_total_memory_qmp = get_polling_output['return']['stats'][keyname]
+            if ((abs(long(guest_memory) - long(stat_total_memory_qmp))) > (long(guest_memory) * ratio)):
```
Comment (xutian) [inline qemu/tests/balloon_service.py]: same as above lines

File: qemu/tests/balloon_service.py [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+                raise error_context.TestFail("total memory of guest %s is not"
+                                             "equal to stat-total-memory in "
```
Comment (xutian) [inline qemu/tests/balloon_service.py]: syntax issue too

File: qemu/tests/balloon_service.py [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+    driver = params.get("driver_name", "balloon")
+
+    error_context.context("Boot guest with balloon device", logging.info)
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=timeout)
+
+    device_path = params.get("device_path", "/machine/peripheral/")
+    device = params.get("balloon", "balloon0")
+    device_path = "%s%s" % (device_path, device)
```
Comment (xutian) [inline qemu/tests/balloon_service.py]: base_path = /machine/peripheral
device = xxx
device_path = "/".join(base_path, device)

**Comment by CongSmile** [comment]
@xutian 

Thanks for the review, could you review it later?

As there are some comments you have told me before have not updated.

File: qemu/tests/balloon_service.py [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+
+        :return drive_letter: virtio-win disk volume letter.
+        """
+        volume_name = params["cdrom_virtiowin"]
+        volume_name = volume_name.split("/")[-1]
+        volume_name = volume_name[:volume_name.rfind(".")]
+        try:
+            return utils_misc.get_win_disk_vol(session,
+                                               key="VolumeName",
+                                               val=volume_name)
```
Comment (suqinhuang) [inline qemu/tests/balloon_service.py]: key and val have been updated to condition

File: qemu/tests/balloon_service.py [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+
+    def get_disk_vol(session):
+        """
+        Get virtio-win disk volume letter.
+
+        :param session: VM session.
+
+        :return drive_letter: virtio-win disk volume letter.
+        """
+        volume_name = params["cdrom_virtiowin"]
```
Comment (suqinhuang) [inline qemu/tests/balloon_service.py]: we always use "cdrom_virtio"

File: qemu/tests/balloon_service.py [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+        """
+        Check / Install balloon service.
+
+        :param session: VM session.
+
+        :return drive_letter: virtio-win disk volume letter.
+        """
+        drive_letter = get_disk_vol(session)
+        status_balloon_service = params["status_balloon_service"] % drive_letter
+        output = session.cmd_output(status_balloon_service)
```
Comment (suqinhuang) [inline qemu/tests/balloon_service.py]: Could you pls add step description

Comment (suqinhuang) [inline qemu/tests/balloon_service.py]: a verb will make the string more clear, e.g set/get/check balloon status

File: qemu/tests/balloon_service.py [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+        if re.search(r"running", output.lower(), re.M):
+            if re.search(r"stop", output.lower(), re.M):
+                logging.debug("Install Balloon Service in guest.")
+                try:
+                    install_balloon_service = params["install_balloon_service"] % drive_letter
+                    session.cmd(install_balloon_service)
+                except ShellCmdError:
+                    raise exceptions.TestError("Install balloon service failed !")
+            else:
+                logging.debug("Run Balloon Service in guest.")
```
Comment (suqinhuang) [inline qemu/tests/balloon_service.py]: Do we still need to run balloon after balloon installation,
if yes, need to remove else ...

File: qemu/tests/balloon_service.py [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+                try:
+                    run_balloon_service = params["run_balloon_service"] % drive_letter
+                    session.cmd(run_balloon_service)
+                except ShellCmdError:
+                    raise exceptions.TestError("Run balloon service failed !")
+        return drive_letter
```
Comment (suqinhuang) [inline qemu/tests/balloon_service.py]: why need to return drive_letter again

**Comment by CongSmile** [comment]
@suqinhuang and @xutian 

The case has been updated, could you please help review it again?

Thanks.

**Comment by CongSmile** [comment]
Test Result:
18:13:07 INFO | PASS smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u1.qcow2.virtio_blk.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.balloon_service

**Comment by suqinhuang** [comment]
Acked-by: Suqin Huang

File: qemu/tests/cfg/balloon_service.cfg [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+    only Windows
+    no Host_RHEL.m5, Host_RHEL.m6
+    monitor_type = "qmp"
+    monitors = qmp1
+    balloon = balloon0
+    balloon_dev_devid = balloon0
+    balloon_dev_add_bus = yes
+    driver_name = "balloon"
+    need_enable_verifier = yes
+    need_clear_verifier = yes
```
Comment (spcui) [inline qemu/tests/cfg/balloon_service.cfg]: not found where use it.

File: qemu/tests/balloon_service.py [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+            guest_mem = mem_base.get_vm_mem(vm)
+
+        stat_memory_qmp = get_polling_output['stats'][keyname]
+        stat_memory_qmp = "%sB" % stat_memory_qmp
+        stat_memory_qmp = int(float(utils_misc.normalize_data_size(
+                                   (stat_memory_qmp), order_magnitude="M")))
+        if (abs(guest_mem - stat_memory_qmp)) > (guest_mem * check_mem_ratio):
+            raise exceptions.TestFail("%s of guest %s is not equal to"
+                                      " %s in qmp, the ratio is %s" %
+                                      keyname, guest_mem, stat_memory_qmp,
```
Comment (spcui) [inline qemu/tests/balloon_service.py]: should be (keyname, guest_mem, stat_memory_qmp, check_mem_ratio)?

**Comment by spcui** [comment]
LGTM,
Acked-by: scui@redhat.com

but this patch deps on MemoryBaseTest(), need wait it.

**Comment by CongSmile** [comment]
Yes, this PR based on MemoryBaseTest(), thanks scui's comments !

**Comment by suqinhuang** [comment]
https://github.com/avocado-framework/avocado-vt/pull/366

**Comment by suqinhuang** [comment]
avocado-framework/avocado-vt#366 have been merged, merged this patch

**Pull request merged**